Meta without dets

mkuratczyk opened the pull request:
Deprecation of ra_log_meta with its DETS-based metadata persistence.
Instead, each server saves its own state in a simple binary file.
Of course for upgradeability, if a file doesn't exist, we take values from
ra_log_meta.

mkuratczyk committed:
New server metadata persistenace

Until now, we persisted metadata for all servers in a given
Ra system in a single DETS file, owned by ra_log_meta gen batch server.
This was a source of bottlenecks, so we move each servers metadata
to a simple binary file. If the file doesn't exist, we try to
fetch metadata from ra_log_meta and immediately persist it, so that
we don't need ra_log_meta going forward.

mkuratczyk committed:
Refactor is_new test

Now that meta_fd is a part of the state, repeated calls
to ra_server:init() lead to different initial states,
so the test is adjusted not to expect the same result
from both calls.

mkuratczyk committed:
Fix ra_server:follower_install_snapshot_machine_version

In this test we end up in update_last_applied with an undefined
meta_fd set in base_state, so let's mock it as a no-op.

gomoripeti commented on src/ra_server_meta.erl:
```diff
@@ -0,0 +1,211 @@
+                          NodeAtomBin = atom_to_binary(NodeAtom, utf8),
+                          NameSize = byte_size(NameAtomBin),
+                          NodeSize = byte_size(NodeAtomBin),
+                          <<NameSize:8/unsigned, NameAtomBin/binary,
```
just noting that although an atom can only contain max 255 (possibly unicode) characters, the binary representation might require more than 255 bytes
```
1> Atom = list_to_atom(lists:duplicate(255, 256)).
'ĀĀĀĀĀĀĀĀĀĀĀĀĀĀĀĀĀĀĀĀĀĀĀĀĀĀĀĀĀĀĀĀĀĀĀĀĀĀĀĀĀĀĀĀĀĀĀĀĀĀĀĀĀĀĀĀĀĀĀĀĀĀĀĀĀĀĀĀĀĀĀĀĀĀĀĀĀĀĀĀĀĀĀĀĀĀĀĀĀĀĀĀĀĀĀĀĀĀĀĀĀĀĀĀĀĀĀĀĀĀĀĀĀĀĀĀĀĀĀĀĀĀĀĀĀĀĀĀĀĀĀĀĀĀĀĀĀĀĀĀĀĀĀĀĀĀĀĀĀĀĀĀĀĀĀĀĀĀĀĀĀĀĀĀĀĀĀĀĀĀĀĀĀĀĀĀĀĀĀĀĀĀĀĀĀĀĀĀĀĀĀĀĀĀĀĀĀĀĀĀĀĀĀĀĀĀĀĀĀĀĀĀĀĀĀĀĀĀĀĀĀĀĀĀĀĀĀĀĀĀĀĀĀĀĀĀĀĀĀĀĀĀĀĀĀĀĀĀĀĀĀĀĀĀĀ'
2> byte_size(atom_to_binary(Atom)).
510
```
term_to_binary stores the atom size in 1 or 2 bytes https://www.erlang.org/doc/apps/erts/erl_ext_dist.html#atom_utf8_ext

sorry if this was considered already, and storing the size in 1 byte is intentional

kjnilsson replied:
Sorry it's late and I clearly cant math. We'd just need 2052 bytes to encode the atoms so we're still well below the block size.

mkuratczyk replied:
Thanks a lot for pointing this out. I've modified the format and code to use 2 bytes for atom sizes. Technically, if both atoms were full of UTF-8, we could still fail, but what are the odds?! :)  Based on our testing, node names can't be UTF-8 (and the would need to be full of them for this to be a problem)

mkuratczyk committed:
Handle UTF-8 in atoms

While an atom cannot be longer than 255 characters, its binary
representation can be longer, if UTF-8 characters are present.

mkuratczyk commented:
Karl is cooking something better
